lib/trie/hash.go

```diff
@@ -5,195 +5,285 @@ package trie
+	type result struct {
+		index  int
+		buffer *bytes.Buffer
+		err    error
+	}
+
+	resultsCh := make(chan result)
+
+	for i, child := range children {
+		go func(index int, child node) {
+			buffer := encodingBufferPool.Get().(*bytes.Buffer)
+			buffer.Reset()
+			// buffer is put back in the pool after processing its
+			// data in the select block below.
+
+			err := encodeChild(child, buffer)
+
+			resultsCh <- result{
+				index:  index,
+				buffer: buffer,
+				err:    err,
 			}
+		}(i, child)
+	}
+
+	currentIndex := 0
+	resultBuffers := make([]*bytes.Buffer, len(children))
+	for range children {
+		result := <-resultsCh
+		if result.err != nil && err == nil { // only set the first error we get
+			err = result.err
 		}
-	} else {
-		for i := 0; i < 16; i++ {
-			if child := b.children[i]; child != nil {
-				scEncChild, err := encodeAndHash(child)
-				if err != nil {
-					return nil, err
-				}
-				h.tmp.Write(scEncChild)
+
+		resultBuffers[result.index] = result.buffer
+
+		// write as many completed buffers to the result buffer.
+		for currentIndex < len(children) &&
+			resultBuffers[currentIndex] != nil {
+			// note buffer.Write copies the byte slice given as argument
+			_, writeErr := buffer.Write(resultBuffers[currentIndex].Bytes())
+			if writeErr != nil && err == nil {
+				err = writeErr
 			}
+
+			encodingBufferPool.Put(resultBuffers[currentIndex])
+			resultBuffers[currentIndex] = nil
+
+			currentIndex++
 		}
 	}
 
-	return h.tmp.Bytes(), nil
+	for _, buffer := range resultBuffers {
+		if buffer == nil { // already emptied and put back in pool
+			continue
+		}
+		encodingBufferPool.Put(buffer)
+	}
+
+	return err
 }
 
-// encodeLeaf encodes a leaf with the encoding specified at the top of this package
-func (h *hasher) encodeLeaf(l *leaf) ([]byte, error) {
-	if !l.dirty && l.encoding != nil {
-		return l.encoding, nil
+func encodeChildsSequentially(children [16]node, buffer *bytes.Buffer) (err error) {
+	for _, child := range children {
+		err = encodeChild(child, buffer)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func encodeChild(child node, buffer *bytes.Buffer) (err error) {
+	if child == nil {
+		return nil
+	}
+
+	scaleEncodedChild, err := encodeAndHash(child)
+	if err != nil {
+		return fmt.Errorf("failed to hash and scale encode child: %w", err)
+	}
+
+	_, err = buffer.Write(scaleEncodedChild)
+	if err != nil {
+		return fmt.Errorf("failed to write child to buffer: %w", err)
 	}
 
-	h.tmp.Reset()
+	return nil
+}
+
+// encodeLeaf encodes a leaf to the buffer given, with the encoding
+// specified at the top of this package.
+func encodeLeaf(l *leaf, buffer *bytes.Buffer) (err error) {
+	l.encodingMu.RLock()
+	if !l.dirty && l.encoding != nil {
+		_, err = buffer.Write(l.encoding)
+		l.encodingMu.RUnlock()
+		if err != nil {
+			return fmt.Errorf("cannot write stored encoding to buffer: %w", err)
+		}
+		return nil
+	}
+	l.encodingMu.RUnlock()
```

Comment:
```suggestion
	l.encodingMu.RLock()
	defer l.encodingMu.RUnlock()
	if !l.dirty && l.encoding != nil {
		_, err = buffer.Write(l.encoding)
		if err != nil {
			return fmt.Errorf("cannot write stored encoding to buffer: %w", err)
		}
		return nil
	}
```

this locking seems simpler, would it be ok?

Reply:
oh wait, i guess you want to unlock before the rest of the function?

Reply:
Yeah I just wanted to release the lock as soon as possible, since it would have to wait for 3 encoding below it before the defer gets executed. But that might be a bit too pre-optimization, let me know I can commit your suggestion.

Reply:
From my understanding of `sync.RWMutex`, it will continue to allow access to read.  If the `if` condition is true then you'll have to wait for for the buffer to write anyways.  I prefer the defer style.